### DISCUSSION TITLE
E3Doolchanger

### DISCUSSION THREAD

**Discussion by TobyCWood** [FormerLurker/Octolapse#838] [General]
I'm trying to see if I can get Octolapse to work with the Toolchanger. Not looking too good at the moment. I got this message...
"Printer Not Supported
You printer does not support a critical gcode, M114, which Octolapse uses to correctly time snapshots. You can solve this problem by flashing your printer with firmware that supports this gcode. I recommend you contact the manufacturer, since this gcode is required for many OctoPrint plugins, and is considered a standard."
Well... I know that RRF does support m114 so... any ideas or is it a no op?

**Comment by TobyCWood** [comment]
M555 P2 didn't work either. :-(

**Reply by FormerLurker** [reply]
I need more info to deal with the problem you are having.  Not only does M114 need to be implemented, but M400 has to be implemented correctly too for it to work.  This is, sadly, a basic limitation of the approach Octolapse uses to stabilize the timelapse.  Some printers will just return the location from M114 immediately, even if the printer still has movements in the planner, which destroys the 'smooth timelapse' effect that Octolapse is known for.  I added this check to the start up (after long deliberation) due to the fact that I was spending a lot of time trying to debug and fix a problem that I've only seen in some uncommon often proprietary firmware.

Anyway, your problem is probably not related to your tool changer.  Octolapse will work fine with a tool changer, as I have one and programmed it to work with one.  It might not work with your firmware, though, if M114 doesn't work as it does in Marlin 1.x-2.x, reprap, smoothieware, and other popular firmware.  If you want to create an issue for this one and fill in all the info (ESPECIALLY logs showing the problem, firmware version with links to code if possible, and your Octolapse settings).  There is a guide on the repo's home page for reporting issues that explains how to get the information I need if you are unsure.  Let me know if you need help finding that.
